### PR TITLE
Add a simple timer support to schedule work at fixed times/intervals

### PR DESCRIPTION
**Pull request by sagar0**
Adding a simple timer support to schedule work at a fixed time.

Test Plan:
Added new unit tests.

### CODE REVIEW

File: util/timer_test.cc [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+
+    ASSERT_EQ(5, count1);
+    ASSERT_EQ(10, count2);
+}
+
+int main(int argc, char** argv) {
+  ::testing::InitGoogleTest(&argc, argv);
+
+  return RUN_ALL_TESTS();
+}
```
Comment (siying) [inline util/timer_test.cc]: Should we start with a new line?

File: util/timer_test.cc [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+  TimerTest()
+      : mock_env_(new ROCKSDB_NAMESPACE::MockTimeEnv(
+          ROCKSDB_NAMESPACE::Env::Default())) {}
+
+ protected:
+  std::unique_ptr<ROCKSDB_NAMESPACE::MockTimeEnv> mock_env_;
+};
+
+TEST_F(TimerTest, SingleScheduleOnceTest) {
+    const uint64_t kSecond = 1000000;  // 1sec = 1000000us
```
Comment (siying) [inline util/timer_test.cc]: Will run "make format" fix this formatting?

File: util/timer_test.cc [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+#include "db/db_test_util.h"
+
+class TimerTest : public testing::Test {
+ public:
+  TimerTest()
+      : mock_env_(new ROCKSDB_NAMESPACE::MockTimeEnv(
+          ROCKSDB_NAMESPACE::Env::Default())) {}
+
+ protected:
+  std::unique_ptr<ROCKSDB_NAMESPACE::MockTimeEnv> mock_env_;
```
Comment (siying) [inline util/timer_test.cc]: Maybe i missed anything but I didn't see feature of MockTimeEnv is used here.
If the purpose is to simulate TimedWait without actual waiting, tests like StatsHistoryTest.RunStatsDumpPeriodSec does something complicated to achieve that.

File: util/timer_test.cc (hunk 2 above)
Comment (siying) [inline util/timer_test.cc]: If the test will wait for seconds to finish, it feels a little bit too long.

**Comment by sagar0** [comment]
Somehow travis builds are no longer getting queued for this PR; not sure why.

**Comment by pdillinger** [comment]
> Somehow travis builds are no longer getting queued for this PR; not sure why.

I think just the link failed to be provided. https://travis-ci.org/github/facebook/rocksdb/builds/667842142

File: util/timer_test.cc [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+
+TEST_F(TimerTest, SingleScheduleOnceTest) {
+  const uint64_t kSecond = 1000000;  // 1sec = 1000000us
+  const int kIterations = 1;
+  uint64_t time_counter = 0;
+  mock_env_->set_current_time(0);
+  ROCKSDB_NAMESPACE::port::Mutex mutex;
+  ROCKSDB_NAMESPACE::port::CondVar test_cv(&mutex);
+
+  ROCKSDB_NAMESPACE::Timer timer(mock_env_.get());
```
Comment (siying) [inline util/timer_test.cc:28]: Nit: the test code is all inside namespace rocksdb, so no need to add "ROCKSDB_NAMESPACE::" which may scare readers with the macro.

**Comment by sagar0** [comment]
Rebased to fix conflicts.

**Comment by sagar0** [comment]
All builds and tests passed. I don't know why AppVeyor is reporting a build failure when I see that all tests passed. 
I am somehow unable to land this PR; @siying could you please help me land this?

**Comment by sagar0** [comment]
Travis CI: https://travis-ci.org/github/facebook/rocksdb/builds/671939724

**Comment by siying** [comment]
@sagar0 can you retry again, or press "Ship It" in https://our.internmc.facebook.com/intern/diff/D20465390/ ?

We have a new workflow now which requires accepting internally too. Now I accepted there and hope it works now.

**Pull request closed without merging**